### ISSUE TITLE
Missing Reflect Padding Layer For Convolution Layer

### ISSUE THREAD

**Issue by thesahibnanda** [keras-team/keras#19343]
## **What is Reflect Padding**
Reflect padding, also referred to as symmetric or mirror padding, is a vital technique in image processing and computer vision tasks. Unlike traditional padding methods, reflect padding extends image borders by mirroring existing pixels, preserving image features and avoiding edge effects during convolutional operations.

The need for reflect padding arises from its ability to maintain image integrity, prevent information loss at boundaries, and ensure accurate analysis in tasks like object detection, image classification, and semantic segmentation.

## **Real Life Applications**
Implementing reflect padding in Keras convolutional layers would enhance the library's functionality and enable users to address these critical issues more effectively. This feature would benefit various real-life applications such as medical imaging, video processing, and image analysis, where maintaining spatial coherence and preserving feature integrity are paramount.

By incorporating reflect padding functionality into Keras, users can expect improved performance and accuracy in their convolutional neural network (CNN) models, facilitating more robust and reliable results in diverse image processing scenarios.

## **How The Code After Implementation Will Look**

**Method 1**
```python
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense

# Define the model
model = Sequential()

# Add a convolutional layer with reflect padding
model.add(Conv2D(filters=32, kernel_size=(3, 3), padding='reflect', activation='relu', input_shape=(28, 28, 1)))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2)))

# Add another convolutional layer with reflect padding
model.add(Conv2D(filters=64, kernel_size=(3, 3), padding='reflect', activation='relu'))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2)))

# Flatten the output before feeding into the fully connected layers
model.add(Flatten())

# Add fully connected layers
model.add(Dense(units=128, activation='relu'))
model.add(Dense(units=10, activation='softmax'))

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Print model summary
model.summary()
```

**Method 2**
```python
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, ReflectionPadding2D  # Import ReflectionPadding2D layer

# Define the model
model = Sequential()

# Add a reflection padding layer
model.add(ReflectionPadding2D(padding=(1, 1), input_shape=(28, 28, 1)))

# Add a convolutional layer without padding
model.add(Conv2D(filters=32, kernel_size=(3, 3), activation='relu'))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2)))

# Add another reflection padding layer
model.add(ReflectionPadding2D(padding=(1, 1)))

# Add a convolutional layer without padding
model.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu'))

# Add a max pooling layer
model.add(MaxPooling2D(pool_size=(2, 2)))

# Flatten the output before feeding into the fully connected layers
model.add(Flatten())

# Add fully connected layers
model.add(Dense(units=128, activation='relu'))
model.add(Dense(units=10, activation='softmax'))

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Print model summary
model.summary()
```

## **How The Implementation Code Will Look (Just An Overview)**
```python
from keras import backend as K
from keras.engine.topology import Layer

class ReflectPadding2D(Layer):
    def __init__(self, padding=(1, 1), **kwargs):
        super(ReflectPadding2D, self).__init__(**kwargs)
        self.padding = padding

    def build(self, input_shape):
        super(ReflectPadding2D, self).build(input_shape)

    def compute_output_shape(self, input_shape):
        return (input_shape[0], input_shape[1] + 2*self.padding[0], input_shape[2] + 2*self.padding[1], input_shape[3])

    def call(self, inputs):
        pattern = [[0, 0],
                   [self.padding[0], self.padding[0]],
                   [self.padding[1], self.padding[1]],
                   [0, 0]]
        padded_inputs = K.pad(inputs, pattern, mode='reflect')
        return padded_inputs

    def get_config(self):
        config = {'padding': self.padding}
        base_config = super(ReflectPadding2D, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))
```
